### ISSUE TITLE
[tx] sequences averaged over padded length

### ISSUE THREAD

**Issue by cwwarren** [NovaSky-AI/SkyRL#467]
*I was poking around skyrl-tx today and noticed this. I saw there was a lot of ongoing work in this area so I'm not sure if this is already known or not.*

---

Short sequences are averaged over the padded length, shrinking both losses and gradients whenever they share a batch with longer samples.

https://github.com/NovaSky-AI/SkyRL/blob/d3133c98a1d2b5d57bc616498b646f120faa260d/skyrl-tx/tx/tinker/engine.py#L231-L232

This can be reproduced by invoking forward_backward for 10 and 100 token inputs alone and as a single batch, observing that the short adapter’s gradient norm in the mixed batch falls to about len_short / len_long of its solo value (~0.1× when pairing 10 tokens with 100).

```
>       assert short_norm_mixed / short_norm_alone == pytest.approx(1.0, rel=1e-1)
E       assert np.float32(0.09607843) == 1.0 ± 0.1
E         
E         comparison failed
E         Obtained: 0.09607843309640884
E         Expected: 1.0 ± 0.1
```

```py
# add to test_engine.py

def tree_l2_norm(tree):
    sq_sum = sum(np.sum(np.square(np.array(x))) for x in jax.tree_util.tree_leaves(tree))
    return np.float32(np.sqrt(sq_sum))


def test_short_sequence_gradients_shrink_in_mixed_batch():
    engine = TinkerEngine(
        base_model_name="Qwen/Qwen3-0.6B",
        checkpoints_base_path="",
        max_lora_adapters=8,
        max_lora_rank=32,
    )

    short_id = "adapter_short"
    long_id = "adapter_long"

    engine.process_single_request(
        types.RequestType.CREATE_MODEL, short_id, {"lora_config": {"rank": 32, "alpha": 32}}
    )
    engine.process_single_request(
        types.RequestType.CREATE_MODEL, long_id, {"lora_config": {"rank": 32, "alpha": 32}}
    )

    short_tokens = list(range(301, 311))
    long_tokens = list(range(401, 501))

    # Baseline: measure gradient magnitude for short sequence processed alone.
    engine.process_forward_backward_batch(
        [(FutureStub(401), short_id, make_fwd_bwd_input([short_tokens]))]
    )
    short_norm_alone = tree_l2_norm(engine.accumulated_grads[short_id])

    engine.accumulated_grads[short_id] = None
    engine.accumulated_grads[long_id] = None

    # Baseline: measure gradient magnitude for long sequence processed alone.
    engine.process_forward_backward_batch(
        [(FutureStub(402), long_id, make_fwd_bwd_input([long_tokens]))]
    )
    long_norm_alone = tree_l2_norm(engine.accumulated_grads[long_id])

    engine.accumulated_grads[short_id] = None
    engine.accumulated_grads[long_id] = None

    # Mixed batch: short and long sequences padded to len(long_tokens).
    engine.process_forward_backward_batch(
        [
            (FutureStub(403), short_id, make_fwd_bwd_input([short_tokens])),
            (FutureStub(404), long_id, make_fwd_bwd_input([long_tokens])),
        ]
    )

    short_norm_mixed = tree_l2_norm(engine.accumulated_grads[short_id])
    long_norm_mixed = tree_l2_norm(engine.accumulated_grads[long_id])

    # Expect the short and long gradient magnitudes to remain intact
    assert short_norm_mixed / short_norm_alone == pytest.approx(1.0, rel=1e-1)
    assert long_norm_mixed / long_norm_alone == pytest.approx(1.0, rel=1e-1)
```

**Comment by cwwarren** [comment]
I believe #462 is the fix for the underlying loss masking issue, looks like it's already known.

**Comment by tyler-griggs** [comment]
Hey @cwwarren , thanks for spotting this! Please let us know if you stumble on any other issues.
